### ISSUE TITLE
"incorrect" locale setting causes malformed URL generation

### ISSUE THREAD

**Issue by sfrsfrsfr** [TwinFan/LiveTraffic#244]
hello,

i was getting these errors when setting up livetraffic with an opensky account:

`LiveTraffic WARN  LTChannel.cpp:610/FetchAllData: OpenSky Live Online: HTTP response is not OK but 400 for https://opensky-network.org/api/states/all?lamin=49,808&lomin=8,188&lamax=50,267&lomax=8,901`

Notice the , used in the coordinates

## Expected behavior
No such error :-)

My Current locale setting: 
`LC_CTYPE="en_US.UTF-8"`
`LC_NUMERIC=de_DE.UTF-8`

After changing LC_NUMERIC to
`export LC_NUMERIC="en_US.UTF-8"`
everything worked as expected

## Technical Info
 - OS: Linux Ubuntu 22.04
 - X-Plane version: 12.0 beta 9
 - LiveTraffic version: 3.1.1

Suggestion: the LC_NUMERIC locale setting should not affect the URL generation used to fetch traffic data.

Thanks for a wonderful plugin!

**Comment by TwinFan** [comment]
Same as [here in the forum](https://forums.x-plane.org/index.php?/forums/topic/273725-computed-opensky-url-wrong-regional-formatting/)  but with actually useful info included, so might be able to work on it.

**Comment by TwinFan** [comment]
[`setlocale()`](https://en.cppreference.com/w/cpp/locale/setlocale) would set the locale globally, not even thread-safe, so it would affect anything running in any parallel X-Plane thread, potentially mid-sprintf… That‘s way too dangerous.

So the solution will be to [query the locale](https://en.cppreference.com/w/cpp/locale/localeconv), and if the `decimal_point` is different from `.` then it would be replaced afterwards in the url string.

**Comment by TwinFan** [comment]
@sfrsfrsfr , can you try the attached version to see if that fixes your problem. I am pretty sure that the correct URL gets produced. I'm just a bit concerned that down the road you might end up with JSON parsing errors as that is what happened during my testing. Not totally unreasonable...the locale will influence interpreting strings as lists and numbers, too. But I'm not on Linux and the way I overruled the locale might not have been proper.

Feedback appreciated!
[LiveTraffic_3.1.2.zip](https://github.com/TwinFan/LiveTraffic/files/9942391/LiveTraffic_3.1.2.zip)

**Comment by sfrsfrsfr** [comment]
Hi TwinFan/LiveTraffic,

i can confirm your concern regarding the JSON parsing errors.

LiveTraffic: Current System time is 2022-11-05 18:28:42Z, current simulated
time is 2022-11-05 18:27:12.0Z
0:01:22,210 LiveTraffic ERROR LTOpenSky.cpp:180/ProcessFetchedData: Parsing
flight data as JSON failed
0:01:22,514 I/PLG: The plugin LiveTraffic 3.1.2 is setting global TCAS
override to 1.
0:01:33,420 E/ATC: Radio transmission required but no voice loaded!
0:01:42,123 LiveTraffic ERROR LTOpenSky.cpp:180/ProcessFetchedData: Parsing
flight data as JSON failed

See also the attached full Log.txt.

My locale settings:

locale
LANG=en_US.UTF-8
LANGUAGE=
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC=de_DE.UTF-8
LC_TIME=de_DE.UTF-8
LC_COLLATE="en_US.UTF-8"
LC_MONETARY=de_DE.UTF-8
LC_MESSAGES="en_US.UTF-8"
LC_PAPER=de_DE.UTF-8
LC_NAME=de_DE.UTF-8
LC_ADDRESS=de_DE.UTF-8
LC_TELEPHONE=de_DE.UTF-8
LC_MEASUREMENT=de_DE.UTF-8
LC_IDENTIFICATION=de_DE.UTF-8
LC_ALL=

Thx for looking into this.
Fyi, setting LC_NUMERIC to us_US.UTF-8 in a small wrapper script works just
fine for me.

Bye, Stefan

**Comment by TwinFan** [comment]
Out of curiosity I've taken a look into the Parson library I use for JSON parsing: It uses many hard-coded characters like `,` for field separation, as per JSON standard, and then `strtod` for parsing numbers, which is locale-dependent, and that's where parsing breaks. But replacing those strings before parsing is a sure path to peril...

So I've looked into thread-local locales once again, and indeed there are options. I've used [this Stackoverflow suggestion](https://stackoverflow.com/a/17173977) now.

My test on Mac (I had a debug statement in the code that set the global(!) locale to "de_DE" when loading the plugin) worked out now. All communication threads where affected, e.g. also reading the weather...

With the attached new version I could finally produce German numbers in the UI while still successfully loading and parsing data (need to watch the decimal comma in the OGN line ;-) )
<img width="673" alt="LT_DE_Locale" src="https://user-images.githubusercontent.com/44291093/200144456-6c72260c-8d40-4054-a366-7443d33309ff.png">

So, @sfrsfrsfr, please try this one:
[LiveTraffic_3.1.2.zip](https://github.com/TwinFan/LiveTraffic/files/9944840/LiveTraffic_3.1.2.zip)

**Comment by sfrsfrsfr** [comment]
Hi TwinFan/LiveTraffic,

this doesn't seem to work on Ubuntu 22.04, X-Plane 12 beta 11.
It again generates OpenSky URLs with , instead of .
Log attached.

Bye, Stefan

**Comment by TwinFan** [comment]
:-(
Unfortunately, attaching files to issues doesn’t work when replying by mail. Can you attach the mailed `Log.txt` file directly on GitHub?

**Comment by sfrsfrsfr** [comment]
sure, it should be available now
[Log.txt.gz](https://github.com/TwinFan/LiveTraffic/files/9945641/Log.txt.gz)

**Comment by TwinFan** [comment]
From your log file I see that your German locale also affects X-Plane system functions:
```
0:00:56,085 W/ATC: Limiting EDGG Area Control Center (118,750, 119,110, 119,150, 1
```
These are frequencies. With English format that would read `118.750, 119.110...`. So your locale does affect -as expected- the entire XP system.

I've now (for this, but also for another Linux issue) set up my own Ubuntu system, installed XP12b11 and LiveTraffic, actually a build even without the "fix" discussed here...and I can still **not** reproduce the issue, though with a system with English language and German regional settings I have pretty exactly the same locale as you:
```
Charlie-lin:~/X-Plane/12$ uname -a
Linux Charlie-lin 5.15.0-52-generic #58-Ubuntu SMP Thu Oct 13 08:03:55 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
Charlie-lin:~/X-Plane/12$ locale
LANG=en_US.UTF-8
LANGUAGE=en_US:en
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC=de_DE.UTF-8
LC_TIME=de_DE.UTF-8
LC_COLLATE="en_US.UTF-8"
LC_MONETARY=de_DE.UTF-8
LC_MESSAGES="en_US.UTF-8"
LC_PAPER=de_DE.UTF-8
LC_NAME=de_DE.UTF-8
LC_ADDRESS=de_DE.UTF-8
LC_TELEPHONE=de_DE.UTF-8
LC_MEASUREMENT=de_DE.UTF-8
LC_IDENTIFICATION=de_DE.UTF-8
LC_ALL=
Charlie-lin:~/X-Plane/12$ ./X-Plane-x86_64 
```
I am not saying the problem doesn't exist...two individuals reported it...but I still can't get there.

**Comment by TwinFan** [comment]
How to you actually start X-Plane? And how does your wrapper script look like.

I start X-Plane from the command line like shown above:
```
Charlie-lin:~/X-Plane/12$ ./X-Plane-x86_64
```
or by double-clicking that executable in a file explorer window.

**Comment by sfrsfrsfr** [comment]
the wrapper script "xplane.sh" is very simple:

```
#!/bin/sh

# 2x FHD
# --monitor_bounds=0,0,1920,1080,1920,0,1920,1080
#
# 1x WQHD, 1x FHD
# --monitor_bounds=0,0,2560,1440,1920,0,1920,1080

CWD=`pwd`
export LC_NUMERIC="en_US.UTF-8"

echo "1 is $1"
echo "2 is $2"
echo "3 is $3"

cd $HOME/X-Plane\ 12
./X-Plane-x86_64 $1 $2 $3

cd $CWD
```

i use i3wm and run the wrapper in a terminal

**Comment by TwinFan** [comment]
A Windows(!) user on XP11(!) now reported something very similar: errors while parsing (here: RealTraffic) tracking data and wrong URLs for requesting data from the internet (here: weather information). See [here](https://forums.x-plane.org/index.php?/forums/topic/273725-regional-formattinglocale-issues-url-json-parsing/&do=findComment&comment=2477715) on the Org.

No feedback as of yet (but hey, it‘s Christmas) if the suggestion to change number formatting on Windows-level helped. But offers the chance for me to try on Windows, which I hadn’t done before.

**Comment by TwinFan** [comment]
[Quote from the support thread](https://forums.x-plane.org/index.php?/forums/topic/273725-regional-formattinglocale-issues-url-json-parsing/&do=findComment&comment=2478367), analysis that I am suspecting _another_ plugin the interfere with the process' locale:

> On a different note: I have compared the log your posted first with the one posted now and I noticed something interesting, which becomes apparent in these lines of the first log from yesterday:
> 
> `0:00:00.000 Altitude/win/XPMP2 INFO  CSLModels.cpp:1091/CSLModelsLoad: Total number of known models now is 5370`
> `...`
> `0:00:50.308 LiveTraffic INFO  LiveTraffic.cpp:420/LoopCBOneTimeSetup: LiveTraffic 3.2.2 (19-Dec-2022) successfully loaded!`
> `0:00:50,308 LiveTraffic INFO  LTApt.cpp:2576/LTAptRefresh: Starting thread to read apt.dat for airports 120,0nm around 82,4999 N / 82,9999 E`
> `0:00:50,308 Altitude/win/XPMP2 DEBUG Map.cpp:338/MapLayerCreate: Created a map layer for map 'XPLM_MAP_IOS'`
> 
> Only look at the timestamp at the very beginning: Both Altitude (which uses the same XPMP2 library as LiveTraffic) and LiveTraffic originally format the timestamp with a dot to separate the fractions of a second from the seconds, but then suddenly, in the mids of processing, that change to a comma!
> 
> I can assure you that neither LiveTraffic nor XPMP2 have any code to influence the regional formatting or the locale.
> 
> My guess is that some other plugin, noting the Spanish regional settings, sets the locale. But if done the straight-forward simple way, then the locale changes for the entire process, i.e. all plugins. It might be worthwhile to change back to Spanish number settings, and then try to remove one of the other plugins one by one to see which one causes the locale change.

**Comment by TwinFan** [comment]
Based on the above idea I created _another_ plugin, with which I can switch the process locale (using [`std::setlocale()`](https://en.cppreference.com/w/cpp/locale/setlocale)) between "C" and "de_DE.UTF-8". And this way I was able to reproduce the issue and can also explain why my [initial fixing attempt above](https://github.com/TwinFan/LiveTraffic/issues/244#issuecomment-1304658910) did not work:

1. X-Plane starts up with a proper "C" locale set process-wide.
2. All plugins start under that "C" locale, too, including all threads of LiveTraffic. Even if for whatever reason the locale is set to some non-C value by the time the plugins start, the the fix above makes sure the "C" locale is correctly set on a thread-bases for the LiveTraffic communication threads. So far, so good.
3. Now some misbehaving plugin comes into play. I am fairly sure in the case of the user "dueronet" in the support forum, the log of @sfrsfrsfr doesn't mention too many plugins, but maybe even there:
    1. This misbehaving plugin sets the process locale to a value taken from the system, probably intending to be nice to the user for some output formatting. But if done wrong, then it affects the locale of the entire process.
    2. With my own misbehaving test plugin I can set the locale at will via menu. When I now set it to "de_DE.UTF-8" process-wide, then again the entire system is affected, **including the threads, which had set a thread-local locale earlier**.
    3. As a result, output is in German format, but all the errors for URL formatting and JSON parsing that users have seen pop up again.

So if any plugin changes the process locale after LiveTraffic's threads have been started, then again that overwrites their local locale and all the effort of setting the thread locale was in vain.

With some effort I could change creating the URLs to modern C++ streams with stream-local locales...but that wouldn't solve the issue with JSON parsing, which happens deep down in another library.

So I could just say: There is another bad behaving plugin, remove that. But maybe I find some other way...and if it is to regularly (re)set the thread-local locale.
